### DISCUSSION TITLE
Error handling with fakeBaseQuery (RTK Query streaming updates)

### DISCUSSION THREAD

**Discussion by morzel85** [reduxjs/redux-toolkit#4993] [Q&A] [answered]
What is the recommended way of handling errors while using streaming updates?
Is there a way to setup RTKQ `createApi` in such a way that errors thrown from `onCacheEntryAdded` or `updateCachedData` affect the `isError` property returned by API hook?

**Comment by seandinwiddie** [comment]
When using RTK Query with streaming updates, error handling can be a bit more complex. Here are some recommendations on how to handle errors:

**1. Use `onError` callback in `fetchBaseQuery`**

You can pass an `onError` callback to `fetchBaseQuery` to catch errors that occur during the fetch process. This callback will be called with the error as an argument.
```js
import { fetchBaseQuery } from '@reduxjs/toolkit/query/react';

const baseQuery = fetchBaseQuery({
  baseUrl: 'https://my-api.com',
  onError: (error) => {
    console.error('Error occurred:', error);
    // You can also dispatch an error action or update the state here
  },
});
```
**2. Use `retry` option in `fetchBaseQuery`**

You can also use the `retry` option to specify a retry policy for failed requests. This can help handle temporary errors or network issues.
```js
const baseQuery = fetchBaseQuery({
  baseUrl: 'https://my-api.com',
  retry: (failureCount, error) => {
    if (failureCount < 3) {
      return true; // retry up to 3 times
    } else {
      return false; // give up after 3 retries
    }
  },
});
```
**3. Catch errors in `onCacheEntryAdded` and `updateCachedData`**

To catch errors thrown from `onCacheEntryAdded` or `updateCachedData`, you can wrap these functions in a `try-catch` block.
```js
import { createApi } from '@reduxjs/toolkit/query/react';

const api = createApi({
  reducerPath:'myApi',
  baseQuery: baseQuery,
  endpoints: (builder) => ({
    getItems: builder.query({
      query: () => 'items',
      onCacheEntryAdded: (cacheEntry, { updateCachedData }) => {
        try {
          // code that might throw an error
        } catch (error) {
          console.error('Error in onCacheEntryAdded:', error);
          // You can also dispatch an error action or update the state here
        }
      },
      updateCachedData: (cacheEntry, { updateResult }) => {
        try {
          // code that might throw an error
        } catch (error) {
          console.error('Error in updateCachedData:', error);
          // You can also dispatch an error action or update the state here
        }
      },
    }),
  }),
});
```
**4. Use `isError` property in API hook**

To affect the `isError` property returned by the API hook, you can throw an error from the `onError` callback or from within the `onCacheEntryAdded` or `updateCachedData` functions. This will set the `isError` property to `true` and provide the error message as the `error` property.
```js
import { useGetItemsQuery } from './api';

const { data, error, isError } = useGetItemsQuery();

if (isError) {
  console.error('Error:', error);
  // Handle error here
}
```
By following these recommendations, you should be able to handle errors effectively when using RTK Query with streaming updates.

**Reply by EskiMojo14** [reply]
chatGPT-tastic

fetchBaseQuery doesn't have onError or retry options, and updateCacheData isn't a valid endpoint key

**Comment by morzel85** [comment]
Somebody asked about it two years earlier: https://github.com/reduxjs/redux-toolkit/discussions/2434
I'll see if I can get some attention to this topic on Discord.

**Reply by markerikson** [reply, marked as answer]
The short answer is that RTK Query is meant to track the status of a promise and cache its result. The lifecycle methods let you kick off additional side effect logic, but essentially anything you do in there is then outside the scope of RTK Query's cache entry.

So, while RTKQ gives you space to write additional logic, any status you would want to track regarding some external connection would need to be handled separately yourself.

Beyond that, we don't currently have a way to force a cache entry to have an errored status.

(for the OP's specific mention of `updateCachedData`, I'm not sure how that factors in here as far as errors.)

**Reply by morzel85** [reply]
Oh, ok. So I'll stay with custom error state for tracking error that happens after stream is established. Although it would be a valuable feature to "natively" support errors in long connections (Websockets, Firestore etc.).
